### ISSUE TITLE
Consider if we can improve the debugging experience with script printers

### ISSUE THREAD

**Issue by ahayzen-kdab** [KDAB/cxx-qt#357]
Currently if you put a breakpoint on an invokable as it's context is C++ the `self` is just a pointer. This makes it hard to inspect the values of the Rust struct and add watches etc.

Consider if we can use script printers to help here at all. Although it's not clear if this could reach through the C++ pointer ->  the m_rustObj -> the Rust struct ?

See some of the Qt ones https://invent.kde.org/kdevelop/kdevelop/-/blob/master/plugins/gdb/printers/qt.py

**Comment by vimpostor** [comment]
Commenting to see if I can assign issues to myself now

**Comment by ahayzen-kdab** [comment]
Note that if the lldb debugger is used we might be able to use the `lldbbridge.py` and `pygdbmi` to help ? (see internal wiki page on VSCode and lldb)

**Comment by vimpostor** [comment]
Yeah I was investigating this quite a bit last time already and found out that the only thing missing is implementing support for pretty printing `std::pin` upstream in Rust itself. For a full list of upstream supported pretty printers see https://github.com/rust-lang/rust/blob/master/src/etc/rust_types.py

The implementation for the various existing pretty printers can be found in that same directory and it should be pretty trivial to hack together a pretty printer for `std::pin` from that (which should be sufficient, because if in the stack trace you go down one frame, where the `std::pin` is dereferenced, the pretty printer will already work).

So it's not a hard issue, and I plan to do it eventually, just didn't have any time for this yet. ;)

**Comment by ahayzen-kdab** [comment]
Note that with Rust 1.71 we might be able to use Debugger visualization attributes to specify the gdb scripts etc for each of the types in cxx-qt-lib, so that the debugging experience would be similar to using QtCreator.

https://blog.rust-lang.org/2023/07/13/Rust-1.71.0.html#debugger-visualization-attributes
